src/approaches/grammar_search_invention_approach.py

```diff
@@ -396,6 +397,220 @@ def _create_grammar(grammar_name: str, dataset: Dataset,
+            score += abs(hadd_heuristic - ideal_heuristic)
+        return score
+
+
+class _HaddLookaheadHeuristic(_HAddBasedHeuristic):
+    """Score predicates by using the hadd values of the induced operators
+    to compute an energy-based policy, and comparing that policy to demos.
+
+    Overview of the idea:
+    1. Predicates induce operators. Denote this ops(preds).
+    2. Operators induce an hadd heuristic. Denote this hadd(state, ops(preds)).
+    3. The heuristic induces a greedy one-step lookahead energy-based policy.
+       Denote this pi(a|s) propto exp(-constant * hadd(succ(s, a), ops(preds)).
+    4. The objective for predicate learning is to maximize prod pi(a | s)
+       where the product is over demonstrations.
+    """
+
+    def _evaluate_atom_trajectory(self, atoms_sequence: List[Set[GroundAtom]],
+                                  hadd_fn: utils.HAddHeuristic,
+                                  ground_ops: Collection[_GroundSTRIPSOperator]
+                                  ) -> float:
+        score = 0
+        for i in range(len(atoms_sequence)-1):
+            atoms, next_atoms = atoms_sequence[i], atoms_sequence[i+1]
+            ground_op_demo_prob = 0. # total probability mass for demo actions
+            ground_op_total_prob = 0.  # total probability mass for all actions
+            for ground_op in ground_ops:
+                # Only care about applicable ground ops.
+                if not ground_op.preconditions.issubset(atoms):
+                    continue
+                # Compute the next state under the operator.
+                successor_atoms = atoms.copy()
+                for atom in ground_op.add_effects:
+                    successor_atoms.add(atom)
+                for atom in ground_op.delete_effects:
+                    successor_atoms.discard(atom)
+                # Compute the heuristic for the successor atoms.
+                h = hadd_fn(utils.atoms_to_tuples(successor_atoms))
+                # Compute the probability that the correct next atoms would be
+                # output under an energy-based policy.
+                k = CFG.grammar_search_lookahead_softmax_constant
+                p = np.exp(-k*h) if not np.isinf(h) else 0.
+                ground_op_total_prob += p
+                # Check whether the successor atoms match the demonstration.
+                if successor_atoms == next_atoms:
+                    ground_op_demo_prob += p
+            # If there is a demonstration state that is a dead-end under the
+            # operators, immediately return a very bad score, because planning
+            # with these operators would never be able to recover the demo.
+            if ground_op_demo_prob == 0:
+                return float("inf")
+            # Accumulate the log probability of each (state, action) in this
+            # demonstrated trajectory.
+            atoms_log_prob = np.log(ground_op_demo_prob) - \
```

Comment:
rename to `trans_log_prob` since it's more about the `atoms -> next_atoms` transition?